### PR TITLE
ADR on import structure 

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

1. Feel free to push commits on this branch!
2. add your name to the deciders if you want to be active part of the discussion 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: docs/adr/0004-import-structure.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+## Pros and Cons of the Options <!-- optional -->
+
+### General import rules 1
+
+- a module should have a clear responsibility,
+   - which should be documented
+   - from that responsibility can be derived its knowledge about the world around it,
+   - which in turn drives which other modules it "sees"
+-  keep __init__ light, as it is always imported when a submodule is imported.
```
Comment (Hugo-Inmanta) [inline docs/adr/0004-import-structure.md:45]:
```suggestion
-  keep `__init__` light, as it is always imported when a submodule is imported.
```

File: docs/adr/0004-import-structure.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  - where only composing modules import the non-interface modules of composed packages.
+  - composing interface modules should also only import interface modules of composed modules
+
+
+e.g. consider executor and deploy packages and the new_agent
+- executor.executor is the interface package
+- deploy imports executor.executor
+- new_agent import both the interface and concrete implementations of both executor and deploy (top level)
+- deploy doesn't import the non-interface parts of executor because it never constructs executors (it is constructed with a reference to a executor manager, which serves as a factory)
+- executor doesn't know about deploy at all
```
Comment (Hugo-Inmanta) [inline docs/adr/0004-import-structure.md:65]: I think some names in this example are outdated which makes it hard to follow

Comment (wouterdb) [reply docs/adr/0004-import-structure.md]: Updated!

File: docs/adr/0004-import-structure.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+## Pros and Cons of the Options <!-- optional -->
+
+### General import rules 1
+
+- a module should have a clear responsibility,
+   - which should be documented
+   - from that responsibility can be derived its knowledge about the world around it,
+   - which in turn drives which other modules it "sees"
+-  keep __init__ light, as it is always imported when a submodule is imported.
+- a package can offer an interface module (called now something like `model`/`types`/`__init__`).....
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply docs/adr/0004-import-structure.md:46]: If you have suggestions?

**Review by arnaudsjs — APPROVED** [review]

File: docs/adr/0004-import-structure.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply docs/adr/0004-import-structure.md:46]: What about `<package_name>_interface` or `<package_name>_api`?
And do we want to encourage using the `__init__.py` file more for this or does it contradict with the rule `keep __init__ light` ?

 Is this more a case by case thing where the interface can live in the `__init__` if it's not too heavy but it should live in its own file as it grows ?

File: docs/adr/0004-import-structure.md [hunk 4]
```diff
@@ -57,12 +57,13 @@ Chosen option: "[option 1]", because [justification. e.g., only option, which me
...
-- new_agent import both the interface and concrete implementations of both executor and deploy (top level)
-- deploy doesn't import the non-interface parts of executor because it never constructs executors (it is constructed with a reference to a executor manager, which serves as a factory)
-- executor doesn't know about deploy at all
+e.g. consider `inmanta.agent.executor.executor` and `inmanta.deploy` packages and the new_agent
+- `inmanta.agent.executor.executor` is the interface package for the executor framework
+- `inmanta.deploy` is the package containing the scheduler (it has no external interface package)
+- `inmanta.deploy` imports `inmanta.agent.executor.executor`
+- `inmanta.agent.agent_new` imports both the interface and concrete implementations of both executor and deploy (top level)
+- `inmanta.deploy` doesn't import the non-interface parts of `inmanta.agent.executor.executor` because it never constructs executors (it is constructed with a reference to a executor manager, which serves as a factory)
+- `inmanta.agent.executor.executor` doesn't know about deploy at all
```
Comment (Hugo-Inmanta) [inline docs/adr/0004-import-structure.md:66]:
```suggestion
e.g. consider `inmanta.agent.executor` and `inmanta.deploy` packages and the new_agent
- `inmanta.agent.executor` is the interface package for the executor framework
- `inmanta.deploy` is the package containing the scheduler (it has no external interface package)
- `inmanta.deploy.scheduler` imports `inmanta.agent.executor`
- `inmanta.agent.agent_new` imports both the interface and concrete implementations of both executor and deploy (top level)
- `inmanta.deploy.scheduler` doesn't import the non-interface parts of `inmanta.agent.executor` because it never constructs executors (They are constructed with a reference to a executor manager, which serves as a factory)
- `inmanta.agent.executor` doesn't know about deploy at all
```

**Review by Hugo-Inmanta — APPROVED** [review]

**Review by jptrindade — APPROVED** [review]
